server/datastore/mysql/vpp.go

```diff
@@ -667,79 +626,243 @@ func (ds *Datastore) InsertVPPToken(ctx context.Context, tok *fleet.VPPTokenData
 	}
 
-	if _, err := ds.writer(ctx).ExecContext(ctx, stmt, teamID, nullTeam, id); err != nil {
-		return ctxerr.Wrap(ctx, err, "updating vpp token team")
+	if err := sqlx.GetContext(ctx, ds.reader(ctx), &tokEnc, stmt, tokenID); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, ctxerr.Wrap(ctx, notFound("VPPToken"), "selecting vpp token from db")
+		}
+		return nil, ctxerr.Wrap(ctx, err, "selecting vpp token from db")
 	}
 
-	return nil
+	tokDec, err := decrypt(tokEnc.Token, ds.serverPrivateKey)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "decrypting vpp token with serverPrivateKey")
+	}
+
+	if err := sqlx.SelectContext(ctx, ds.reader(ctx), &tokTeams, stmtTeams, tokenID); err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "selecting vpp token teams from db")
+	}
+
+	tok := &fleet.VPPTokenDB{
+		ID:        tokEnc.ID,
+		OrgName:   tokEnc.OrgName,
+		Location:  tokEnc.Location,
+		RenewDate: tokEnc.RenewDate,
+		Token:     string(tokDec),
+	}
+
+	if tokTeams == nil {
+		// Not assigned, no need to loop over teams
+		return tok, nil
+	}
+
+TEAMLOOP:
+	for _, team := range tokTeams {
+		switch team.NullTeam {
+		case fleet.NullTeamAllTeams:
+			// This should only be possible if there are no other teams
+			// Make sure something array is non-nil
+			tok.Teams = []fleet.TeamTuple{}
+			break TEAMLOOP
+		case fleet.NullTeamNoTeam:
+			tok.Teams = append(tok.Teams, fleet.TeamTuple{
+				ID:   0,
+				Name: fleet.TeamNameNoTeam,
+			})
+		case fleet.NullTeamNone:
+			// Regular team
+			tok.Teams = append(tok.Teams, fleet.TeamTuple{
+				ID:   *team.TeamID,
+				Name: team.Name,
+			})
+		}
+	}
+
+	return tok, nil
+}
+
+func (ds *Datastore) UpdateVPPTokenTeams(ctx context.Context, id uint, teams []uint) (*fleet.VPPTokenDB, error) {
+	stmtRemove := `DELETE FROM vpp_token_teams WHERE vpp_token_id = ?`
+	stmtInsert := `
+	INSERT INTO
+		vpp_token_teams (
+			vpp_token_id,
+			team_id,
+			null_team_type
+	) VALUES `
+	stmtValues := `(?, ?, ?)`
+
+	var values string
+	var args []any
+	// No DB constraint for null_team_type, if no team or all teams
+	// comes up we have to check it in go
+	var nullTeamCheck fleet.NullTeamType
+
+	if len(teams) > 0 {
+		for _, team := range teams {
+			team := team
+			if values == "" {
+				values = stmtValues
+			} else {
+				values = strings.Join([]string{values, stmtValues}, ",")
+			}
+			var teamptr *uint
+			nullTeam := fleet.NullTeamNone
+			if team != 0 {
+				// Regular team
+				teamptr = &team
+			} else {
+				// NoTeam team
+				nullTeam = fleet.NullTeamNoTeam
+				nullTeamCheck = fleet.NullTeamNoTeam
+			}
+			args = append(args, id, teamptr, nullTeam)
+		}
+	} else if teams != nil {
+		// Empty but not nil, All Teams!
+		values = stmtValues
+		args = append(args, id, nil, fleet.NullTeamAllTeams)
+		nullTeamCheck = fleet.NullTeamAllTeams
+	}
+
+	stmtInsertFull := stmtInsert + values
+
+	if err := ds.checkVPPNullTeam(ctx, &id, nullTeamCheck); err != nil {
```

Comment:
The problem here is that we check outside the transaction, in the reader. This is racy and we could insert inconsistent data. I think we could make that bullet-proof by doing the check at the same time as the INSERT, i.e. (untested, just winging it):

```
INSERT INTO
		vpp_token_teams (
			vpp_token_id,
			team_id,
			null_team_type
	) SELECT ?, ?, ? FROM DUAL WHERE NOT EXISTS (SELECT 1 FROM vpp_token_teams WHERE <all teams>)
```

or similar if we try to insert a "no team" (there shouldn't be another "no team" or another "all teams"). We do something like that to make the MDM profile names unique across Windows and Apple (which are in two different tables, so we can't do it with a DB constraint): https://github.com/fleetdm/fleet/blob/1f5c7792a903e0302e8eb469f4db022a0de657d3/server/datastore/mysql/microsoft_mdm.go#L1532-L1543

Reply:
`SELECT ?, ?, ? FROM DUAL` Very esoteric 👽

Reply:
Does an `INSERT INTO ... SELECT` return an error if the `WHERE NOT EXISTS` fails? It seems it would just not insert anything instead of returning an error?

Reply:
Correct, it needs an additional check to see if it did insert anything, as we do here for profiles: https://github.com/fleetdm/fleet/blob/1f5c7792a903e0302e8eb469f4db022a0de657d3/server/datastore/mysql/microsoft_mdm.go#L1565-L1572

Reply:
Doing the simpler thing for now and checking using the same method, but inside the transaction, as per our slack conversation